### DISCUSSION TITLE
SAML Login - Any Easy Method? Working

### DISCUSSION THREAD

**Discussion by manu-remsense** [cypress-io/cypress#28384] [Questions and Help] [answered]
Hey guys,

I'm using Cognito for identity federation, which uses Microsoft as IdP with SAML type of authentication. I've been having massive issues in trying to get E2E test with authentication to work.

My web app basically checks the cookies of the request, and if the request doesn't contain valid cookies, it redirects to AWS cognito endpoint, which then redirects to a Microsoft login. Upon successful login, it then redirects back to the app.

I've tried using this guide,
https://docs.cypress.io/guides/end-to-end-testing/azure-active-directory-authentication

Which doesn't work due to the login.microsoftonline.com page not loading. The error states, 
`Failed to find a valid digest in the 'integrity' attribute for resource '<URL>' with computed SHA-384 integrity '<sha-384 code here>'. The resource has been blocked.`
As a result it causes the following error, 
`Uncaught [Retry 0] Failed to load external resource ['https://aadcdn.msftauth.net/shared/1.0/content/js/ConvergedLogin_PCore_LRl_HOP41vTvA9IKhSqNsg2.js'], reloading from fallback CDN endpoint`

Then I decided to implement login flow programmatically, from my research it seems that SAML authentication is hard to set up without the browser. I've briefly tried just now, but not too successful yet, it seems a bit challenging as I have to reverse-engineer how microsoft's login works. 

Is there any easy way of setting this up? Anyone had any success with SAML authentication with Cypress? I've tried looking all over online, but nothing useful so far. Any help would be appreciated..

Thanks,
Manu

**Comment by portyu9** [comment, marked as answer]
Hi @manu-remsense 

The specific error you posted:

```text
Failed to find a valid digest in the 'integrity' attribute ...
The resource has been blocked.
```

is now covered directly by Cypress's Azure AD documentation.

Cypress rewrites some HTML/JavaScript so that third-party authentication pages can run inside the Cypress architecture. That changes the resource bytes, which can invalidate Subresource Integrity (SRI) hashes.

On current Cypress, configure:

```js
const { defineConfig } = require('cypress')

module.exports = defineConfig({
  removeSRIAttributes: true,

  e2e: {
    experimentalModifyObstructiveThirdPartyCode: true,
  },
})
```

`removeSRIAttributes` is the current supported option for stripping SRI attributes that would otherwise block rewritten resources.

Then model the Microsoft portion with `cy.origin()`:

```js
cy.visit('/')

cy.get('[data-cy="login"]').click()

cy.origin(
  'https://login.microsoftonline.com',
  { args: { username } },
  ({ username }) => {
    cy.get('input[type="email"]').type(username, {
      log: false,
    })

    cy.get('input[type="submit"]').click()
  }
)
```

Depending on the Microsoft account, the flow can move to another origin such as `login.live.com`; if so, use another `cy.origin()` block for that origin.

After successful authentication:

```js
cy.location('origin').should('eq', 'https://your-app.example.com')
```

and cache the authenticated result with `cy.session()` so you do not drive the entire SAML/AAD UI before every test.

So I would **not** reverse-engineer Microsoft's SAML protocol manually unless UI authentication genuinely cannot be used. Your original SRI error has a Cypress-supported workaround now.

Please mark this answer as accepted if it helped, thank you.

**Reply by manu-remsense** [reply]
Thanks for the reply, as it is an old post, I managed to solve it by getting Cognito credentials with a test account which meant that the Microsoft login wasn't needed with cypress. Hope this answer can help another another person who needs to auth via Microsoft login, cheers!

Manu
